x-pack/plugins/notifications/server/service/action_result.js

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+/**
+ * Action Results represent generic, predictable responses from Actions.
+ */
+export class ActionResult {
+
+  /**
+   * Create a new Action Result.
+   *
+   * Success is determined by the existence of an error.
+   *
+   * @param {String} message The message to display about the result, presumably in a Toast.
+   * @param {Object|undefined} response The response from the "other" side.
+   * @param {Object|undefined} error The error, if any.
+   */
+  constructor({ message, response, error }) {
+    this.message = message;
+    this.response = response;
+    this.error = error;
+    this.ok = !Boolean(error);
+  }
+
+  /**
+   * Get the error caused by the action.
+   *
+   * @returns {Object|undefined} The error response, or {@code undefined} if no error.
+   */
+  getError() {
```

Comment:
This class looks so clean!  I like that you expose the inner variables as functions rather than leaving them to be directly accessed.  That was something I was going back and forth on with my panel actions/embeddable stuff.  Any particular reason you like this approach best?  I don't think we are very consistent about it.  Just curious.

Reply:
Thanks!

I have been working with Typescript recently and it feels more appropriate to have a defined object rather than a bunch of arbitrary properties for a contract like this where some things are inferred. It also enables any future `Action`s to extend `ActionResult` rather than trying to figure out how to make a custom object that matches all of the same fields and their behavior (e.g., overriding `isOk`).

In terms of this code, even I am somewhat inconsistent. The missing fields response is a plain Javascript `Object` with a predictable structure. I thought about making a class out of it, but it doesn't need one because nothing is inferred or even dynamic about it -- it just needs a strict interface definition, which Typescript should bring us in pretty short order. Hopefully we'll develop better collective habits once we are able to embrace Typescript within the codebase.

Reply:
You can still override properties without making getters (this.isOk = mycustomFunc in the super class), and I think you can still use a typescript interface to define variables as well as functions expected.

But last night I thought of a reason to prefer using getter functions - to give a place where you could put deprecation warnings if you ever wanted to get rid of some variables!  At least that works well for a class that acts an a pluggable API.